[Dasco Gabriel] iP

DystoriaX opened the pull request:
# Duke
Your personal task manager. :exploding_head:

- Text-based
- _Intuitive_
- Suitable for people who ~is lazy~ **type fast**

Get started with 4 steps:
1. Download the program [here](https://github.com/DystoriaX/ip/releases/download/A-Jar/main.jar)
2. Open your CLI and navigate to the folder containing the `.jar` file
3. Run `java -jar main.jar [path_to_file]`, where `path_to_file` is the path of the file that you want to store your task into.

That's all!

Features:
- [x] Manage task
- [x] Find task
- [ ] Reminder

> To be or not to be
> ~ you-know-who

```java
public class Main{
    public static void main(String[] args) {
        System.out.println("Hello world\n");
    }
}
```

Commit:
Add Gradle support

DystoriaX committed:
Add greet, echo, and exit functionality

DystoriaX committed:
Duke class: Change static methods to instance-level methods

DystoriaX committed:
Implement list functionality

DystoriaX committed:
Add Task and TaskList classes

DystoriaX committed:
Add initialization for `done` field

DystoriaX committed:
Implement mark done functionality

DystoriaX committed:
Change toString representation for Task and TaskList class

DystoriaX committed:
Create task package

DystoriaX committed:
Add Todo, Deadline, and Event classes

DystoriaX committed:
Add text UI test

DystoriaX committed:
Add exception handling with DukeException class

DystoriaX committed:
Add exception handling test cases

DystoriaX committed:
Implement delete task functionality

DystoriaX committed:
Add delete task test cases

DystoriaX committed:
Implement save feature

DystoriaX committed:
Implement date using java.time.LocalDate

DystoriaX committed:
Merge branch 'branch-level-7'

DystoriaX committed:
Erase unnecessary try-catch block in addEvent method

DystoriaX committed:
Merge branch 'branch-level-8'

DystoriaX committed:
Fix bug for invalid format after merging

DystoriaX committed:
Implement Storage class

DystoriaX committed:
Duke class: remove unnecessary imports

DystoriaX committed:
Add Ui class

DystoriaX committed:
Move Storage and Ui classes to their own packages

DystoriaX committed:
Add Parser and Command class

DystoriaX committed:
Move all classes to duke package

DystoriaX committed:
Add unit test for task package

DystoriaX committed:
Fix minor UI mistake

DystoriaX committed:
Add flexibility to choose location to store data

DystoriaX committed:
Fix minor UI mistake

DystoriaX committed:
Add Javadoc to public methods

DystoriaX committed:
Remove TODO comment

DystoriaX committed:
Parser class: Add private access modifier for filePath

DystoriaX committed:
task package: Change boolean field done to isDone

DystoriaX committed:
ui package: Add JavaDoc

DystoriaX committed:
Add find feature

DystoriaX committed:
Merge branch 'branch-A-JavaDoc'

DystoriaX committed:
Merge branch 'branch-A-CodingStandard'

DystoriaX committed:
Merge branch 'branch-Level-9'

benedictchuajj commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,62 @@
+
+public class Duke {
+    private Ui ui;
+    private TaskList taskList;
```
Small nit but perhaps use a plural form name such as *tasks* to represent a collection of objects?

benedictchuajj commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,62 @@
+import duke.task.TaskList;
+import duke.ui.Ui;
+
+public class Duke {
```
Perhaps add a descriptive header comment for Duke and its methods since they are public?
> Write descriptive header comments for all public classes/methods. - [CS2103T Java coding standard](https://se-education.org/guides/conventions/java/intermediate.html#comments)

DystoriaX replied:
Ah yes. Thanks for the catch!

benedictchuajj commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,31 @@
+        this.task = task;
+    }
+
+    @Override
+    public void execute(TaskList taskList, Ui ui, Storage storage) throws IOException {
+        taskList.addTask(this.task);
+        ui.sayAddTask(this.task, taskList.size());
+        storage.store(taskList.serialize());
+    }
```
Since there's no Javadoc from the inherited class for execute, perhaps you can add a Javadoc here explaining what execute will do. I noticed the same issue in a few other classes that inherit from Command.

benedictchuajj commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,63 @@
+        } else if (input.startsWith("delete ")) {
+            int index = Integer.parseInt(input.substring(7));
+            return new DeleteCommand(index);
+        } else if (input.startsWith("todo ")) {
```
Perhaps you can use switch statements here to improve readability for this method. 😃

benedictchuajj commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,43 @@
+
+/** A class that represents a deadline task */
+public class Deadline extends Task {
+    private LocalDate by;
```
Small nit here, but perhaps rename *by* to *byDate* for greater clarity that it is a LocalDate object?

benedictchuajj commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,107 @@
+        this.isDone = false;
+    }
+
+    /**
+     * Serializes the task as a string representation.
+     *
+     * @return Serialized string representation.
+     */
+    public String serialize() {
+        return this.description;
+    }
```
Any reason why this is a separate method from getDescription instead of just reusing getDescription? If they serve the same intended purpose, perhaps you can remove one of them?

DystoriaX replied:
`getDescription()` is intended to access the `description` property from outside. I think implementation wise, I'd still prefer to directly call `this.description` as there is no abstraction barrier that we break. I guess this also helps improve the program performance by a bit.

benedictchuajj commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,107 @@
+                String at = params[3];
+
+                return new Event(description, at, done);
+            }
```
Perhaps you can consider using switch statements here too.

benedictchuajj reviewed (commented):
Looks good with very few coding standard violations. 👍
Just a few nits to fix to improve readability.

Keep up the good work for iP 😃

ailing35 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,62 @@
+                Command c = Parser.parse(response);
+                c.execute(taskList, ui, storage);
+                isExit = c.isExit();
+            } catch (IOException e) {
+                ui.sayError(e.getMessage());
+            } catch (DukeException e) {
+                ui.sayError(e.getMessage());
+            }
```
Perhaps these two catch statements can be merged into one using ` ||` since they have the same body?

DystoriaX replied:
That's a good idea. Thanks!

ailing35 commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,31 @@
+package duke.command;
+
+import duke.storage.Storage;
+
+import duke.task.Task;
+import duke.task.TaskList;
+
+import duke.ui.Ui;
```
Perhaps it's better to group these together for better consistency across different files?

ailing35 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,79 @@
+     * @throws IOException If an I/O error occurs.
+     */
+    public void store(String data) throws IOException {
+        FileWriter fout = new FileWriter(filePath.toString());
+        BufferedWriter bout = new BufferedWriter(fout);
```
Could the namings be better? One suggestion is using `bufferedOutput` instead of `bout`, or since the scope of use for the variable is small, it could be `b`.

ailing35 commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,111 @@
+
+/** A class that manages the UI */
+public class Ui {
+    public static final String LOGO = " ____        _        \n"
+            + "|  _ \\ _   _| | _____ \n"
+            + "| | | | | | | |/ / _ \\\n"
+            + "| |_| | |_| |   <  __/\n"
+            + "|____/ \\__,_|_|\\_\\___|\n";
+    public static final String MESSAGE_GREETING = "Hello from\n" + LOGO;
+    public static final String MESSAGE_HELP = "How can I help you?";
+    public static final String MESSAGE_ADD_TASK = "Added the following task:\n"
+            + "  %1$s\n"
+            + "You have %2$d task(s) in the list";
+    public static final String MESSAGE_DELETE_TASK = "Deleted the following task:\n"
+            + "  %1$s\n"
+            + "You have %2$d task(s) in the list";
+    public static final String MESSAGE_LIST_TASKS = "Your current task(s):\n"
+            + "%1$s";
+    public static final String MESSAGE_FOUND_TASKS = "Found the following task(s):\n"
+            + "%1$s";
+    public static final String MESSAGE_MARK_TASK = "Marked task %1$d as done";
+    public static final String MESSAGE_LOADING_SUCCESS = "Data loaded";
+    public static final String MESSAGE_LOADING_FAIL = "Failed to load data";
+    public static final String MESSAGE_GOODBYE = "Good bye!";
```
Any reason why these are all `public` instead of `private`?

DystoriaX replied:
Since it's constant, I guess it may be a good idea if other classes can access it. Outside user won't be able to modify it either, so this is to improve the flexibility. Will consider to change it to `private` if it is necessary. Thanks!

ailing35 reviewed (commented):
Overall great use of OOP, perhaps can add more javadoc comments for consistency.

DystoriaX committed:
Duke class: Add comment header and simplify catch

DystoriaX committed:
Merge branch 'add-gradle-support'

DystoriaX committed:
Add checkstyle to gradle

DystoriaX committed:
Merge branch 'branch-A-CheckStyle'

DystoriaX committed:
Add GUI

DystoriaX committed:
Merge branch 'branch-Level-10'

DystoriaX committed:
command package: Handle out of bound errors

Delete command and mark command does not handle error when user tries
to access invalid index.

This affects the user experience, where the app will crash / continue
to run without any further notice of what errors happened in the app.

Let's throw DukeException when `execute` method is called on both
commands, where the exception contains the message that user tries
to access invalid index.

DystoriaX committed:
Add necessary assertions

DystoriaX committed:
task package: Remove unused methods

DystoriaX committed:
Refactor Parser class

The parse method in Parser class seems to be not readable. It lacks of
necessary abstraction.

Let's:
- use switch instead of if-else statements
- use enums for different types of command
- extract methods to parse each command in order to comply with SLAP

Doing so will improve the code readability. It helps to ensure that
the code is safe as we use enums here.

DystoriaX committed:
Merge pull request #1 from DystoriaX/branch-A-Assertions

Branch a assertions

DystoriaX committed:
Merge branch 'master' of https://github.com/DystoriaX/ip into branch-A-CodeQuality

DystoriaX committed:
Merge pull request #2 from DystoriaX/branch-A-CodeQuality

Branch a code quality

DystoriaX committed:
Create main.yml

DystoriaX committed:
Add priority feature

DystoriaX committed:
task package: Fix errors and update unit tests

DystoriaX committed:
Fix code style issues detected by checkstyle

DystoriaX committed:
Merge pull request #3 from DystoriaX/branch-BCD-Extension

Add priority feature

DystoriaX committed:
Improve GUI

DystoriaX committed:
Add user guide

DystoriaX committed:
Add product image

DystoriaX committed:
Merge pull request #5 from DystoriaX/branch-A-UserGuide

Create User Guide

DystoriaX committed:
Update typo in User Guide

DystoriaX committed:
Set theme jekyll-theme-slate
